### ISSUE TITLE
Misunderstanding of ollama num_ctx parameter and context window

### ISSUE THREAD

**Issue by PhilipAmadasun** [ollama/ollama#2714]
I'm trying to understand the relationship between the context window and the `num_ctx` parameter.  Let's say I'm using mistral, and mistral's max context (according to google) is 8000, and "attention span" (according to google) is 128000. If I have a 27000 length user query. What exactly happens? If I set `num_ctx: 4096`. Does mistral just grab the last 4096 token sequence from the 27K user query? Then process the 4096 sequence along with the 128K window it grabs from the previously established overall context (In the case of the RESTful API, I'm talking about that body['context'] thing)?

**Comment by jmorganca** [comment]
Hi there,

Two things happen,

1. If you are using the Chat API, it will only send as many messages as can fit in the context window.
2. If it's still too big (e.g. a huge user message), then the prompt will roughly be split in half, opening up another 1/2 of the context window for new token generations (and it will continue doing this as tokens are generated)

There's a lot of work to do to improve this further - would love any feedback

Hope this helps

**Comment by PhilipAmadasun** [comment]
@jmorganca So if user query is 27K tokens, and mistrals max context length is 8K. The 27K will be be split to 14K and then to 7K? If so, then we have 4 sets of 7K tokens. Then Each set goes in as input to the model one at a time? I'm sorry for my confusion, if possible please use numbers in your explanation so maybe it can be clearer to me.

Just to make sure, when you say "context window" do you mean "attention span"? As in how much of the previous query and answer pairs the model can take in for context? 

Or do you means "context window" as in maximum amount of tokens from the current user query that the model can take in as input?

**Comment by luc99hen** [comment]
I also has this question. There is ambiguity between `num_ctx` for ollama and `context window` for a model. Or in other words, could you give some advice on how to set this parameter for a new model? @jmorganca

**Comment by fgenie** [comment]
Why is this closed w/o any conclusion?

**Comment by mitar** [comment]
We also had to debug this for days. Until we found a hidden `num_ctx` in model documentation (not API documentation) which was artificially lowering the allowed context from model's 8k to 2k (default), making outputs really bad for our 4k prompt. Why is `num_ctx` not set to model's max context by default?

**Comment by FellowTraveler** [comment]
I think I have to create a custom modelfile whenever I want to be able to load up any model with a different num_ctx than the default of 2048.
This leaves me wondering if Ollama is making a copy of every model that I use in this way, or if it just references the original model without having to copy it on the hard drive.
Also it seems strange that I have to create a custom modelfile at all, and otherwise be stuck with a 2048 context window for every single model by default.

**Comment by mitar** [comment]
You can provide `num_ctx` in the API call.

**Comment by andreashappe** [comment]
> You can provide `num_ctx` in the API call.

I tried this yesterday, but when using llama3 the used token count (according to the HTTP response) always stayed below 2k

**Comment by mitar** [comment]
But is your prompt larger than 2k?

**Comment by andreashappe** [comment]
> But is your prompt larger than 2k? Does your response require more than 2k?

Yes, I am using it within https://github.com/ipa-lab/hackingBuddyGPT to compare llama3 and openai LLMs and when looking at the stats reported by the HTTP response I can see that openai uses 8k tokens for the request while llama3 always caps at <= 2k, even when setting the `num_ctx` parameter.

**Comment by mitar** [comment]
Maybe it is a bug in the tool you are using.

**Comment by andreashappe** [comment]
quites sure it's not (I've written that tool on my own).

If I am using openAI API directly, it uses (depending upon the use-case) 8-100k context size.

When I start llama3 with ollama and use its OpenAI-compatible API (and add the options -> num_ctx parameter, setting it to 4096 or 8192 does not matter) and keep all other things identical -> used context size is hard limited to 2k. I am using the token counts reported by the ollama openai-compatible API, so I am not counting them myself.

So I am quite sure that it is not the tool itself..

**Comment by MarkoSagadin** [comment]
@andreashappe I think I know what might be your issue with the token counting.

If you are using OpenAI-compatible API, that means that your text passes through OpenAI SDK logic, gets to the Ollama server, which generates a response and passes that back again through the OpenAI SDK logic.

The number of input and generated tokens that you see in the response object is calculated by the OpenAI SDK, not by Ollama.

How can OpenAI SDK know what kind of model are you using to use the correct tokenizer? I think it can't, it just uses TikToken library to calculate number of tokens in strings.

The only way to count the tokens going in or out of Ollama (AFAIK) is with the Huggingface's Tokenizer class in the Transformers library.

**Comment by andreashappe** [comment]
@MarkoSagadin With API I meant the [ollama openAI-compatible API](https://github.com/ollama/ollama/blob/main/docs/api.md) with the model set to 'llama3' (so that it will use the correct one). I myself am only using the HTTP interface with a direct HTTP call (through the python requests library). I never mentioned the SDK.

And the token count is the count that is returned by ollama's HTTP response, so I am quite sure that it knows what it is dealing with. Or am I getting the architecture totally wrong (which can always be the case)?

**Comment by MarkoSagadin** [comment]
Aha, I thought that you were talking about this [OpenAi compatible API](https://github.com/ollama/ollama/blob/main/docs/openai.md).

I agree with you, the Ollama should know how to count tokens.

I need to check how values in `prompt_eval_count` and `eval_count` compare against Groq.

**Comment by MarkoSagadin** [comment]
They compare badly. 

Below table is from the tool that I linked earlier.

![image](https://github.com/ollama/ollama/assets/41839945/acded338-5c6c-4a7d-a987-661f10b0918e)

* "Max context" is the context window size.
* "Combined" is combined token count of system message and chat history (a list of user/assistant messages) 
* Each column after the “Combined” column shows the number of tokens for the combined content of system message, chat history and the input file that was tested. The values for Ollama models are using `prompt_eval_count` values.

For the first file the there is only one token of difference between the Groq and Ollama models.
After that the difference significantly increases.

Something is off with token counting...

**Comment by andreashappe** [comment]
@MarkoSagadin I am also trying this locally right now (using llama3-8b).. I am using the following answer fields (for the token counts): response['usage']['prompt_tokens'] and response['usage']['completion_tokens'].

I do log away the prompts and the responding answer fields. When I look into the output, then I see the following:

- prompt: 2526 characters -> 436 request tokens, 21 completion tokens
- prompt: 25325 characters -> 1660 request tokens, 98 completion tokens
- prompt: 26090 characters -> 203 request tokens, 12 completion tokens

I am not sure why a larger prompt now creates a smaller request token count :-/ this is with the default 2048 context size.

When I now add the [num_ctx option](https://github.com/ollama/ollama/blob/main/docs/faq.md):

- prompt: 2399 characters -> 62 request tokens, 16 completion tokens
- prompt: 25764 characters -> 1977 request tokens, 153 completion tokens (this is why I believe that there is a 2k limit in place)
- prompt: 25772 characters -> 1980 request tokens, 113 completion tokens

So in the first run I do not understand how the request token count goes down while the prompt itself is more or less the same. In the second run, the request token count keeps high, but seems to be cut-off at roughly 2k.

I am not sure if this is a token counting problem or if I have setup the num_ctx parameter wrong (or if the setting is ignored)

**Comment by mitar** [comment]
Possibly related: https://github.com/ollama/ollama/issues/3427

**Comment by andreashappe** [comment]
okay, I just switched the api url from ollama (localhost:10434) to groq (and removed the `options` from the http request headers) and the context size looks as expected (grows to around 8000). So I am quite sure that it might be something with the ollama http openai api.

**Comment by andreashappe** [comment]
> Possibly related: #3427

Good point. I think this might explain my first testrun (where the token count gets lower).. but would this also explain the second run where the context size seems to hit 2k and then is never increased (to the 8k that I am passing as http request header). Can I check somehow what is used as context size? maybe i am setting it wrong (but I don't get any error -.-)

**Comment by FellowTraveler** [comment]
> You can provide `num_ctx` in the API call.

Thank you, I found the global setting after you said this. I was looking for a way to configure it per-model, but this is better than nothing for sure.

I'm still curious if it actually makes a separate copy of the model weights when I "copy" a model to make a custom modelfile. Hopefully not.

**Comment by itsPreto** [comment]
> Why is num_ctx not set to model's max context by default?

?

**Comment by anrgct** [comment]
When using open-webui, I've noticed that long contextual messages sent to ollama consistently result in poor responses. After investigating the issue, it appears that the `/api/chat` and `/v1/chat/completions` endpoints are defaulting to a 1k context limit. This means that when the content exceeds this length, the system automatically discards the earlier portions, leading to subpar answers. What follows is the captured network request data for open-webui version 0.3.8. 
```
curl 'http://localhost:11434/api/chat' \
-X POST \
-H 'Host: localhost:11434' \
-H 'Accept: */*' \
-H 'User-Agent: Python/3.11 aiohttp/3.9.5' \
-H 'Content-Type: text/plain; charset=utf-8' \
--data-raw '{"model": "qwen1_5-4b-chat-q4_k_m", "messages": [{"role": "user", "content": "<long context>"}], "options": {}, "stream": true}' 
```

**Comment by anrgct** [comment]
I opened a new issue about this bug. @chris-31337 https://github.com/ollama/ollama/issues/6026

**Comment by shenhai-ran** [comment]
I found under each model's `model` page, there is something called `context_length` (see below image). I am wondering, is this JUST a piece of information listed here, or is it a kind of parameter related to the model?

![image](https://github.com/user-attachments/assets/96ee51a6-fab6-4cce-9241-511c83203df5)


Thanks!

**Comment by neel6762** [comment]
> 2\. If it's still too big (e.g. a huge user message), then the prompt will roughly be split in half, opening up another 1/2 of the context window for new token generations (and it will continue doing this as tokens are generated)

@jmorganca I noticed this, while playing with a larger input. For instance, if I set the num_ctx to 50_000 and the input message exceeds this length, the chat method only reads 1/2 of the tokens. Is there a way to compute the # of tokens before passing it to the model? This would simply mean using the tokenizer of the model (but it's not accessible via Ollama) to get the size of the  input tokens, before passing it to the model. OR am I missing something here ..
